Interop/StubGenerator/src/main/kotlin/org/jetbrains/kotlin/native/interop/gen/StubIrTextEmitter.kt

```diff
@@ -0,0 +1,670 @@
+
+        out(";")
+        out("")
+        block("companion object") {
+            aliasConstants.forEach {
+                val mainConstant = canonicalsByValue[it.constant.value]!!
+                out("val ${it.name.asSimpleName()} = ${mainConstant.name.asSimpleName()}")
+            }
+            if (aliasConstants.isNotEmpty()) out("")
+
+            out("fun byValue(value: $baseKotlinType) = " +
+                    "$simpleKotlinName.values().find { it.value == value }!!")
+        }
+        out("")
+        block("class Var(rawPtr: NativePtr) : CEnumVar(rawPtr)") {
+            out("companion object : Type($basePointedTypeName.size.toInt())")
+            out("var value: $simpleKotlinName")
+            out("    get() = byValue(this.reinterpret<$basePointedTypeName>().value)")
+            out("    set(value) { this.reinterpret<$basePointedTypeName>().value = value.value }")
+        }
+    }
+
+    private fun renderFunctionReceiver(receiver: ReceiverParameterStub): String {
+        return renderStubType(receiver.type)
+    }
+
+    private fun renderFunctionParameter(parameter: FunctionParameterStub): String {
+        val annotations = if (parameter.annotations.isEmpty())
+            ""
+        else
+            parameter.annotations.joinToString { renderAnnotation(it) } + " "
+        val vararg = if (parameter.isVararg) "vararg " else ""
+        return "$annotations$vararg${parameter.name.asSimpleName()}: ${renderStubType(parameter.type)}"
+    }
+
+    private fun renderMemberModality(modality: MemberStubModality): String = when (modality) {
+        MemberStubModality.OVERRIDE -> "override "
+        MemberStubModality.OPEN -> "open "
+        MemberStubModality.NONE -> ""
+        MemberStubModality.FINAL -> "final "
+    }
+
+    private fun renderVisibilityModifier(visibilityModifier: VisibilityModifier) = when (visibilityModifier) {
+        VisibilityModifier.PRIVATE -> "private "
+        VisibilityModifier.PROTECTED -> "protected "
+        VisibilityModifier.INTERNAL -> "internal "
+        VisibilityModifier.PUBLIC -> ""
+    }
+
+    private fun renderClassHeader(classStub: ClassStub): String {
+        val modality = when (classStub) {
+            is ClassStub.Simple -> renderClassStubModality(classStub.modality)
+            is ClassStub.Companion -> ""
+            is ClassStub.Enum -> "enum class "
+        }.let { it }
+        val className = when (classStub) {
+            is ClassStub.Simple -> renderClassifierDeclaration(classStub.classifier)
+            is ClassStub.Companion -> "companion object"
+            is ClassStub.Enum -> renderClassifierDeclaration(classStub.classifier)
+        }
+        val constructorParams = when (classStub) {
+            is ClassStub.Simple -> renderConstructorParams(classStub.constructorParams)
+            is ClassStub.Companion -> ""
+            is ClassStub.Enum -> renderConstructorParams(classStub.constructorParams)
+        }
+
+        val inheritance = run {
```

Comment:
The code in this block is rather complicated.
I believe it can be made much simpler by extracting list of supers first (including super class and interfaces), maybe as mutable list to avoid sophisticated concatenations.